File: base-v3/src/main/java/com/adyen/checkout/base/model/paymentmethods/PaymentMethod.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 27/5/2019.
 */

package com.adyen.checkout.base.model.paymentmethods;

import android.os.Parcel;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.adyen.checkout.core.exception.ModelSerializationException;
import com.adyen.checkout.core.model.JsonUtils;
import com.adyen.checkout.core.model.ModelObject;
import com.adyen.checkout.core.model.ModelUtils;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

@SuppressWarnings({"MemberName", "PMD.DataClass"})
public class PaymentMethod extends ModelObject {
    @NonNull
    public static final Creator<PaymentMethod> CREATOR = new Creator<>(PaymentMethod.class);

    private static final String CONFIGURATION = "configuration";
    private static final String DETAILS = "details";
    private static final String GROUP = "group";
    private static final String NAME = "name";
    private static final String BRANDS = "brands";
    private static final String PAYMENT_METHOD_DATA = "paymentMethodData";
    private static final String SUPPORTS_RECURRING = "supportsRecurring";
    private static final String TYPE = "type";

    @NonNull
    public static final Serializer<PaymentMethod> SERIALIZER = new Serializer<PaymentMethod>() {

        @Override
        @NonNull
        public JSONObject serialize(@NonNull PaymentMethod modelObject) {
            final JSONObject jsonObject = new JSONObject();
            try {
                jsonObject.putOpt(CONFIGURATION, modelObject.getConfiguration());
                jsonObject.putOpt(DETAILS, ModelUtils.serializeOptList(modelObject.getDetails(), InputDetail.SERIALIZER));
                jsonObject.putOpt(GROUP, ModelUtils.serializeOpt(modelObject.getGroup(), Group.SERIALIZER));
                jsonObject.putOpt(NAME, modelObject.getName());
                jsonObject.putOpt(BRANDS, JsonUtils.serializeOptStringList(modelObject.getBrands()));
                jsonObject.putOpt(PAYMENT_METHOD_DATA, modelObject.getPaymentMethodData());
                jsonObject.putOpt(SUPPORTS_RECURRING, modelObject.getSupportsRecurring());
                jsonObject.putOpt(TYPE, modelObject.getType());
            } catch (JSONException e) {
                throw new ModelSerializationException(PaymentMethod.class, e);
            }
            return jsonObject;
        }

        @Override
        @NonNull
        public PaymentMethod deserialize(@NonNull JSONObject jsonObject) {
            final PaymentMethod paymentMethod = new PaymentMethod();
            paymentMethod.setConfiguration(jsonObject.optString(CONFIGURATION, null));
            paymentMethod.setDetails(ModelUtils.deserializeOptList(jsonObject.optJSONArray(DETAILS), InputDetail.SERIALIZER));
            paymentMethod.setGroup(ModelUtils.deserializeOpt(jsonObject.optJSONObject(GROUP), Group.SERIALIZER));
            paymentMethod.setName(jsonObject.optString(NAME, null));
            paymentMethod.setBrands(JsonUtils.parseOptStringList(jsonObject.optJSONArray(BRANDS)));
            paymentMethod.setPaymentMethodData(jsonObject.optString(PAYMENT_METHOD_DATA, null));
            paymentMethod.setSupportsRecurring(jsonObject.optBoolean(SUPPORTS_RECURRING, false));
            paymentMethod.setType(jsonObject.optString(TYPE, null));
            return paymentMethod;
        }
    };

    // TODO: 15/04/2019 how to handle the configuration??
    // Configuration is a generic data object that can change per payment method. Save raw string to deserialize later as needed.
    private String configuration;
    private List<InputDetail> details;
    private Group group;
    private String name;
    private List<String> brands;
    private String paymentMethodData;
    private boolean supportsRecurring;
    private String type;

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this));
    }

    @Nullable
    public String getConfiguration() {
        return configuration;
    }

    @Nullable
    public List<InputDetail> getDetails() {
        return details;
    }

    @Nullable
    public Group getGroup() {
        return group;
    }

    @Nullable
    public String getName() {
        return name;
    }

    @Nullable
    public String getPaymentMethodData() {
        return paymentMethodData;
    }

    public boolean getSupportsRecurring() {
        return supportsRecurring;
    }

    @Nullable
    public String getType() {
        return type;
    }

    public void setConfiguration(@Nullable String configuration) {
        this.configuration = configuration;
    }

    public void setDetails(@Nullable List<InputDetail> details) {
        this.details = details;
    }

    public void setGroup(@Nullable Group group) {
        this.group = group;
    }

    public void setName(@Nullable String name) {
        this.name = name;
    }

    public void setBrands(@Nullable List<String> brands) {
        this.brands = brands;
    }

    @Nullable
    public List<String> getBrands() {
        return brands;
    }

    public void setPaymentMethodData(@Nullable String paymentMethodData) {
        this.paymentMethodData = paymentMethodData;
    }

    public void setSupportsRecurring(boolean supportsRecurring) {
        this.supportsRecurring = supportsRecurring;
    }

    public void setType(@Nullable String type) {
        this.type = type;
    }
}

File: components-core/src/main/java/com/adyen/checkout/components/core/PaymentMethod.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 27/5/2019.
 */
package com.adyen.checkout.components.core

import com.adyen.checkout.core.old.exception.ModelSerializationException
import com.adyen.checkout.core.old.internal.data.model.JsonUtils.parseOptStringList
import com.adyen.checkout.core.old.internal.data.model.JsonUtils.serializeOptStringList
import com.adyen.checkout.core.old.internal.data.model.ModelObject
import com.adyen.checkout.core.old.internal.data.model.ModelUtils.deserializeOpt
import com.adyen.checkout.core.old.internal.data.model.ModelUtils.deserializeOptList
import com.adyen.checkout.core.old.internal.data.model.ModelUtils.serializeOpt
import com.adyen.checkout.core.old.internal.data.model.ModelUtils.serializeOptList
import com.adyen.checkout.core.old.internal.data.model.getStringOrNull
import kotlinx.parcelize.Parcelize
import org.json.JSONException
import org.json.JSONObject

@Parcelize
data class PaymentMethod(
    var type: String? = null,
    var name: String? = null,
    var brands: List<String>? = null,
    var brand: String? = null,
    var fundingSource: String? = null,
    var issuers: List<Issuer>? = null,
    var configuration: Configuration? = null,
    var details: List<InputDetail>? = null,
    var apps: List<AppData>? = null,
) : ModelObject() {

    companion object {
        private const val TYPE = "type"
        private const val NAME = "name"

        // Brands is only used for type = scheme
        private const val BRANDS = "brands"

        // Brand is only used for type = giftcard
        private const val BRAND = "brand"
        private const val FUNDING_SOURCE = "fundingSource"
        private const val ISSUERS = "issuers"
        private const val APPS = "apps"
        private const val CONFIGURATION = "configuration"

        // This field is returned in older API versions, only used to retrieve the issuers list
        private const val DETAILS = "details"

        @JvmField
        val SERIALIZER: Serializer<PaymentMethod> = object : Serializer<PaymentMethod> {
            override fun serialize(modelObject: PaymentMethod): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(TYPE, modelObject.type)
                        putOpt(NAME, modelObject.name)
                        putOpt(BRANDS, serializeOptStringList(modelObject.brands))
                        putOpt(BRAND, modelObject.brand)
                        putOpt(FUNDING_SOURCE, modelObject.fundingSource)
                        putOpt(ISSUERS, serializeOptList(modelObject.issuers, Issuer.SERIALIZER))
                        putOpt(CONFIGURATION, serializeOpt(modelObject.configuration, Configuration.SERIALIZER))
                        putOpt(DETAILS, serializeOptList(modelObject.details, InputDetail.SERIALIZER))
                        putOpt(APPS, serializeOptList(modelObject.apps, AppData.SERIALIZER))
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(PaymentMethod::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject): PaymentMethod {
                return PaymentMethod(
                    type = jsonObject.getStringOrNull(TYPE),
                    name = jsonObject.getStringOrNull(NAME),
                    brands = parseOptStringList(jsonObject.optJSONArray(BRANDS)),
                    brand = jsonObject.getStringOrNull(BRAND),
                    fundingSource = jsonObject.getStringOrNull(FUNDING_SOURCE),
                    issuers = deserializeOptList(jsonObject.optJSONArray(ISSUERS), Issuer.SERIALIZER),
                    configuration = deserializeOpt(jsonObject.optJSONObject(CONFIGURATION), Configuration.SERIALIZER),
                    details = deserializeOptList(jsonObject.optJSONArray(DETAILS), InputDetail.SERIALIZER),
                    apps = deserializeOptList(jsonObject.optJSONArray(APPS), AppData.SERIALIZER),
                )
            }
        }
    }
}
